parsl/usage_tracking/usage.py

```diff
@@ -98,90 +99,121 @@ def __init__(self, dfk, port=50077,
         self.domain_name = domain_name
         # The sock timeout will only apply to UDP send and not domain resolution
         self.sock_timeout = 5
-        self.UDP_PORT = port
+        self.udp_port = port
         self.procs = []
         self.dfk = dfk
         self.config = self.dfk.config
-        self.uuid = str(uuid.uuid4())
+        self.correlator_uuid = str(uuid.uuid4())
         self.parsl_version = PARSL_VERSION
-        self.python_version = "{}.{}.{}".format(sys.version_info.major,
-                                                sys.version_info.minor,
-                                                sys.version_info.micro)
-        self.tracking_enabled = self.check_tracking_enabled()
-        logger.debug("Tracking status: {}".format(self.tracking_enabled))
+        self.python_version = (
+            f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
+        )
+        self.tracking_level = self.check_tracking_level()
+        self.start_time = None
+        logger.debug(f"Tracking level: {self.tracking_level}")
+
+    def check_tracking_level(self):
+        """Check if tracking is enabled and return level.
+
+        Returns: int
+            - 0 : Tracking is disabled
+            - 1 : Tracking is enabled with level 1
+                  Share info about Parsl version, Python version, platform
+            - 2 : Tracking is enabled with level 2
+                  Share info about config + level 1
+            - 3 : Tracking is enabled with level 3
+                  Share info about app count, app fails, execution time + level 2
+        """
 
-    def check_tracking_enabled(self):
-        """Check if tracking is enabled.
+        level = 0
 
-        Tracking will be enabled unless either of these is true:
+        if self.config.usage_tracking:
+            level = 1 if self.config.usage_tracking is True else int(self.config.usage_tracking)
```

Comment:
I think this code is expecting that self.config.usage_tracking might be a general int. `Config` should be constraining self.config.usage_tracking to always be a `bool`.

Reply:
Yes. Config does constrain usage_tracking to be a bool. I can change it to be a an int, and set it within `Config` as
```python
class Config(...):
  def __init__(self, ..., usage_tracking: int = 0, ... ):
    ...
    self.usage_tracking = int(usage_tracking) # we change currently set True or False to be Level 1 and 0
    ...
```

Or we can add another Parameter ```usage_tracking_level: int``` in `Config`

What do you think?

Reply:
better to have one parameter that contains the whole config.

you could maybe validate its range within `Config.__init__` too, to raise an exception as early as possible.

Reply:
this is also an indication that another programmer working on the code might make similar mistakes in future, so it would be a good thing to add some kind of test case in parsl/tests/ - for example, you could add a test that checks that Config objects can be instantiated with some appropriate usage tracking values (True, False, 0, 3), and fail with some inappropriate ones ("yes", 4, -1)